### DISCUSSION TITLE
How to manage local state in Rust?

### DISCUSSION THREAD

**Discussion by AlienKevin** [fzyzcjy/flutter_rust_bridge#251] [Q&A] [answered]
## Question
I have a large and complex local state that is only needed in Rust and is costly to send back and forth. What's the best way to manage the local state in Rust?

## Attempt
My current attempt is as follows:
In `api.rs`, I have a struct called `Api` that contains most of my local state:
```
struct Api {
    pub dict: RichDict,
    pub release_time: DateTime<Utc>,
}
```
I wrapped the two states in a struct so I can use Serde to serialize and deserialize.
Then, I implemented some associated functions to initialize and update the `Api` struct, along with some methods to retrieve information from the struct:
```
impl Api {
    pub fn new(app_dir: String) -> Self {
        // Load/create the Cantonese dictionary and set the release_time
    }

    pub fn pr_search(&self, capacity: u32, query: &str) -> Vec<PrSearchResult> {
        // search for a pronunciation in dictionary definitions
    }

    fn get_new_dict<P: AsRef<Path>>(api_path: &P) -> Api {
        // Download the latest dictionary from server
    }
}
```
Next, I create a thread_local `RefCell` to store the `Api` struct for use throughout the app session in Rust:
```
thread_local!(static API: RefCell<Option<Api>> = RefCell::new(None));
```
I then create a `init_api` function that can be called from Dart to initialize the `Api` struct in Rust:
```
pub fn init_api(input_app_dir: String) -> Result<()> {
    API.with(|api_cell| {
        *api_cell.borrow_mut() = Some(Api::new(input_app_dir));
    });
    Ok(())
}
```
Lastly, I implement a public function called `pr_search` to query the local `Api` struct:
```
pub fn pr_search(capacity: u32, query: String) -> Result<Vec<PrSearchResult>> {
    API.with(|api_cell|
        Ok(api_cell.borrow().as_ref().unwrap().pr_search(capacity, &query))
    )
}
```

## Issues encountered
1. Receive ` FfiException(PANIC_ERROR, Once instance has previously been poisoned, null)` after **hot restart**. Presumably the `RefCell` is declared/used twice??
2. `println!()` does not work inside associated functions and methods of the `Api` struct, so it's hard for me to debug them

**Comment by fzyzcjy** [comment, marked as answer]
Btw you can put it to Issues section as well ;)

Personally, I use a "pool". https://github.com/fzyzcjy/flutter_rust_bridge/issues/241

Or, if this https://github.com/fzyzcjy/flutter_rust_bridge/issues/243 is implemented, everything is solved! (You are welcomed to make a PR and I am willing to review and give suggestions!)

> Receive FfiException(PANIC_ERROR, Once instance has previously been poisoned, null) after hot restart. Presumably the RefCell is declared/used twice??

Flutter hot reload will re-run all *dart* code from the beginning. But it has no magic so no *Rust* code will be reset/reexecuted/etc.

**Reply by AlienKevin** [reply]
The "pool" idea seems very interesting. Though my use case here might even be simpler as Dart doesn't need to touch the resource at all. Do you have ideas on why the `println!` are not working inside the methods? Second, I don't understand why the once isntance is poisoned. Is thread_local `RefCell` not a good choice here?

**Reply by fzyzcjy** [reply]
> Do you have ideas on why the println! are not working inside the methods?

Does it even work **OUTSIDE**?
e.g. for android, you need some android logging interface (find on on crates.io). println does NOT work for android.

> , I don't understand why the once isntance is poisoned. Is thread_local RefCell not a good choice here?

not sure.
btw, be careful: flutter_rust_bridge by default (you can customize) use a thread pool to execute rust code. so thread_local may not work.
indeed parking lot's mutex does not have very high cost, so you may want that

**Reply by AlienKevin** [reply]
> e.g. for android, you need some android logging interface (find on on crates.io).

I see. To use println on iOS, is the additional setup available on crates.io as well? (I did a rough search but can't find any)

> indeed parking lot's mutex does not have very high cost, so you may want that

I'll give mutex a spin, thanks!

**Reply by AlienKevin** [reply]
For my final solutions, see issue #252

**Comment by jawadkalia** [comment]
Can mutexes be replaced with RwLocks?

**Reply by fzyzcjy** [reply]
Why not :)

**Reply by jawadkalia** [reply]
Well i kinda tried it. with `mutex` and `thread_local!` and it didnt work. although it did work with `lazy_static!` and `Mutex` and it worked as well for `RwLock`

i wanted to implement per struct level `thread_local!` but im still learning and flailing around. therefore i wasnt able to implement it

**Comment by patmuk** [comment]
Hi,

allow me to extend this question:
I plan to store the complete state of the app in the rust code, having flutter only as a stateless UI layer (or: just with UI state, but not the app logic state).

Though at the start of the app, the state needs to be restored and at the end it should be saved. 
How can I archive this best? I am fine to wait for V2 - so if there are differences, please answer with these in mind!

Here is how I implemented it for now (following the above solution):
```
// holds the complete state of the app, as a global static variable
#[derive(Default)]
#[frb(non_final)]
struct AppState {
    pub model: TodoListModel,
}
// initializes the app_state only at first call
// The app state is behind a mutex to avoid race conditions

lazy_static! {
    static ref APP_STATE: RwLock<AppState> = RwLock::new(AppState::default());
}
            
pub fn view() -> ViewModel {    
    todo_list::view(&APP_STATE.read().model)
}
```
(you can find my complete project in https://github.com/patmuk/flutter-rust-bridge_crux_style/tree/flutter-rust-bridge_codegen)

I have a few questions (and assumptions for the answers):
- should I use this struct, or rather an RustOpaque type?
  - zero-copy (which I pan to use next) is without overhead for Rust -> Dart, but I assume there is overhead in the opposite direction (Dart -> Rust)? Is that serialization & copy overhead?at would be, in my case, at the start of the app (loading the state)
  - Am I right that this would not be the case with a RustOpaque - but that would not be readable by rust? Can I imagine this just as a pointer to (for Dart) binary data? Can I persist that from the Dart side, so I can load it at app start?

- concurrency:
  - the solution above says to use lazy_static! WITH thread_local (I am using the default thread pool). But I don't know how to combine these two macros - and in the example code I did not saw thread_local used at all. Is lazy_static already creating a thread-aware singleton?
  - I assume the RwLock is needed, as the generated Dart functions are all async. Thus I assume that, if two write calls are executed from Dart, they can have a race condition. Or is this handled in FRB?
  - the manual states that complex structs needs it's custom types boxed (in the example above I should use 
```
struct AppState {
    pub model: Box<TodoListModel>,
    // pub model: TodoListModel,
}

```
- is that outdated? (As this seems to work without the box. I can create a PR fixing this in the documentation).

Many thanks for reading through my long questions and the answers in advance!

**Reply by patmuk** [reply]
Oh, I have an additional question: I assumed that I need to persist the struct, which keeps the apps state, through Flutter.
But maybe somebody knows how to do this within rust, for iOS and Android?

**Reply by fzyzcjy** [reply]
I may come back to this later, but for a quick answer: In v2, there is "rust AUTO opaque",  which means you can do something like:

```rust
#[frb(opaque)]
pub struct MyState {
  whatever heavy state here, it will not be serialized and deserialized to the Dart side
}

impl MyState {
  // you can have `&mut WhateverType`, `&WhateverType`, `WhateverType`, just follow the standard rust semantics
  fn update_something(&mut self, ...) { }
}

**Reply by fzyzcjy** [reply]
Btw, I will have a dedicated page in `how-to` section of V2 doc discussing states; and possibly also make a simple demo in the v2 gallery.

**Reply by patmuk** [reply]
Great! I you need help, you can count on me :)

**Reply by fzyzcjy** [reply]
Well... I find out that I cannot say a lot about this - Just use the new RustAutoOpaque feature and write standard Rust code, and that's all (or at least that should be all, before I come across problems - if any - when doing the demo). 😆

![image](https://github.com/fzyzcjy/flutter_rust_bridge/assets/5236035/0436acbd-8dec-443a-84e2-50a1d0311fbe)

**Reply by patmuk** [reply]
Thanks - this is good to know as well. I assume that you don’t persist the state over app restarts?
I will update when I figure this.
In essence, I just need a strict on the rust side (and opaque sounds promising (KISS :) ), and a set/get call to store/load the state (if that needs to be done one the Dart side).

**Reply by fzyzcjy** [reply]
> I assume that you don’t persist the state over app restarts?

Sure. But if you want, it is pretty easy. For example, use `serde` to convert your state to a string (or vec u8), and write to a file on disk.

**Reply by patmuk** [reply]
Ah, great, will do that!
So, an opaque struct and reading/writing to  disk from rust seems like the best solution.

Did I understood correctly that concurrency has to be handled? 
Looking into your implementation of the default handler, it spawns threads using the threadpool crate's execute functions (which uses the std::thread (?) and not, e.g., tokio (?)) for each generated (wire) function call in rust, so that this function's usage of a (std::)mutex/rwLock would work (and is needed).
Without a rust call from Dart, which modifies the state-struct, could be called concurrently multiple times, leading to a data race, right?

**Reply by fzyzcjy** [reply]
> Did I understood correctly that concurrency has to be handled?

Not on your side, FRB handles it :)
Your code just use the most intuitive style, no any extra boilerplate

**Reply by patmuk** [reply]
Great! Btw: Many thanks for the fast replies, very much appreciated!

I would like to understand this better - and the implications (but feel free to ignore my questions, I don't want to steel too much of your time!).

The mutex handling happens here (right?):
```
#[cfg(not(wasm))]
mod io {
    use super::*;
    use lazy_static::lazy_static;
    use parking_lot::Mutex;
    use threadpool::ThreadPool;

    lazy_static! {
        pub static ref THREAD_POOL: Mutex<ThreadPool> = Mutex::new(ThreadPool::with_name(
            "frb_workerpool".into(),
            get_worker_count()
        ));
    }
}
```
and 
```
#[macro_export]
macro_rules! spawn {
    ($($tt:tt)*) => {{
        let worker = $crate::transfer!($($tt)*);
        #[cfg(not(target_family = "wasm"))]
        {
            $crate::thread::THREAD_POOL.lock().execute(worker)
        }
```

Thus threads in the thread pool are never running concurrently, because whenever a thread is spawned, it needs to gain the lock.
I understand that this is not a limitation, as within this one thread, one could spawn more threads (and handle data races and the like on one's own). 

How is this interacting from the dart side? Is a rust functions called in a new isolate? Though the global Thread-pool lock would execute them in sequence.

To use an example, say we have `fn function_A` and `fn function_B`, calling it from the generated Dart function would execute them in parallel to dart's main isolate, but in sequence, as they are synchronized on the mutex on the thread pool (which is the point I probably did not get correct?). However, if in, say, function_B threads are spawn, they have to wait for function_A to release the lock (if it got it first), but can then run in parallel (up to the user to handle concurrency).

This is a very nice implementation - both taking the worry about concurrency on the rust side from the user, providing calls from the dart side as futures, so they don't block it, while still enabling multithreaded programming on rust without influencing/limiting it!

The part I don't understand is that there is a thread-pool, but because of the mutex the threads never run in parallel.
What did I get wrong? If this is nothing, it is still a nice implementation :)

**Reply by fzyzcjy** [reply]
You are welcome!

> The mutex handling happens here

Are you talking about v1? The v2 has a refactored codebase, so many things are neater (I hope so!). I personally suggest wait a bit and check v2 codebase directly :)

> Thus threads in the thread pool are never running concurrently, because whenever a thread is spawned, it needs to gain the lock.

No, `execute` happens very quickly, it is indeed "submit-the-work-and-return".

**Reply by patmuk** [reply]
Oh, yes, I looked into the code of release 1.84 ... will wait for v2! (Is the code in the branch https://github.com/fzyzcjy/flutter_rust_bridge/tree/feat/11073, or split across multiple branches? Though I can wait for the release (and contribute, if you see how I could! (documentation, testing?)).

Ah, I get it now :) The mutex on the pool prevents a race condition, where the pool would concurrently provide the same thread ... so the lock is released after execute returned a thread, and not hold for the threads execution ... Rust's locking mechanism is not that easy to get :)

However, I don't understand how data races are prevented, if not handled buy FBR's user? But I take your word for it and will look in the new code in v2 to understand that later. No need to explain me soon-to-be obsolete code!

**Reply by fzyzcjy** [reply]
> Is the code in the branch

Yes, but a bit messy now, and heavily missing documentation :)

> and contribute, if you see how I could

Welcome!

> However, I don't understand how data races are prevented, if not handled buy FBR's user? But I take your word for it and will look in the new code in v2 to understand that later. No need to explain me soon-to-be obsolete code!

Sure, if v2 still confuses you, feel free to re-ask!

**Reply by patmuk** [reply]
Will do - great, many thanks!!!

**Reply by fzyzcjy** [reply]
You are welcome!

**Comment by AlienKevin** [comment]
A followup question after a long time: I started to see **WatchDogTermination errors** in production a while back and recently I narrowed down the reason to my heavy JSON parsing code in the initialization of the rust Api. I have this standard code in Dart to initialize the Api and I think the **`MyAppApiImpl(dylib)` is blocking the main isolate**, causing the UI to freeze for 2 seconds at the start of the app. Here's my `ffi.dart` which initializes the global `api`:
```dart
const base = 'my_app_api';
final path = Platform.isWindows ? '$base.dll' : 'lib$base.so';
late final dylib = loadLibForFlutter(path);
late final api = MyAppApiImpl(dylib);
```

I saw the example code in V2 and it has this neat async `RustLib.init` function:
```
Future<void> main() async {
  await RustLib.init();
  runApp(const MyApp());
}
```
I think if I wrap the `RustLib.init()` call in a `Future`, that will unblock the main thread right?

On the other hand, it will take some time before I can migrate the whole codebase to V2, so I'm interested in a solution in V1 as well. 
My current understanding is that the Rust code runs in its own thread but the main Dart isolate is blocked because it needs to wait until the Rust code returns. The only workaround I can think of is to invoke the `MyAppApiImpl(dylib)` in its own isolate. However, the [FRB doc](https://cjycode.com/flutter_rust_bridge/feature/async_dart.html?highlight=isolate#asynchronous-in-dart) explicitly discourages the usage of isolates:
> Remark: A common mistake is to call Rust code in another Dart isolate (i.e. "thread") instead of the main isolate. That is completely not needed, and will only make your life harder. As is described above, even if your Rust code computes for 100ms, the async call will only take, say, 0.1ms, and will not block your UI.

The above remark is true for most api calls but I don't think it applies to the Api constructor. Still, I'm not sure if spawning a new isolate is the best approach here.

**Reply by fzyzcjy** [reply]
In v2, the suggested way to do init is:

![image](https://github.com/fzyzcjy/flutter_rust_bridge/assets/5236035/bb85f7ff-4027-4561-a7d6-80317ab56eed)

(the doc is on master, but I have not configured the deployment logic yet)

> On the other hand, it will take some time before I can migrate the whole codebase to V2, so I'm interested in a solution in V1 as well.
My current understanding is that the Rust code runs in its own thread but the main Dart isolate is blocked because it needs to wait until the Rust code returns. The only workaround I can think of is to invoke the MyAppApiImpl(dylib) in its own isolate. However, the [FRB doc](https://cjycode.com/flutter_rust_bridge/feature/async_dart.html?highlight=isolate#asynchronous-in-dart) explicitly discourages the usage of isolates:

What about doing similar things in v1? i.e. Do not do setup inside MyAppApiImpl constructor, but instead do it inside a separate api.whateverFunction();

**Reply by AlienKevin** [reply]
> What about doing similar things in v1? i.e. Do not do setup inside MyAppApiImpl constructor, but instead do it inside a separate api.whateverFunction();

I think this could work but it might be slightly inconvenient: in addition to having to unwrap the `Future<Api>` on the Dart side, I also need to unwrap a `Mutex` on the Rust side. My `Api` is an immutable global variable on the Rust side:
```
lazy_static! {
    static ref API: Arc<Api> = Arc::new(Api::new());
}
```
If I use a separate `Api::init` method, I will have to make this global variable mutable, meaning now I need to wrap the whole api in a `Mutex` and have to unwrap the mutex on each access.

**Reply by fzyzcjy** [reply]
Hmm, what about using a RustOpaque in v1 to hold your API?

**Reply by patmuk** [reply]
@AlienKevin, on the flutter side you don't have to call your constructor - it is called when the lazy_static reference is called for the first time.
So, if you do that later, that should run the initialization code after your app started.  
However, as you mentioned, it should not block anyways. I probably did not get the complete picture.  

Maybe it helps showing my current implementation (using a mutex - but it should work with a mutable reference as well. I did not implemented RustOpaque for now, as I had conflicts with RwLock - so I postponed that for v2).

You can find my code [here](https://github.com/patmuk/flutter-rust-bridge_crux_style/tree/flutter-rust-bridge_codegen), but let me show you the important bits:

(from app_core/src/api.rs):
```
use once_cell::sync::Lazy;
use parking_lot::RwLock;

/// instanciate the lazy static app state -
/// call if you want to control when the app state is initialized,
/// which might take time (due to IO when loading the last saved state)
/// otherwise it is called automatically when the lazy reference is accessed the first time
pub fn init() {
  let _ = &*API;
}

// NOTE that I am using "use once_cell::sync::Lazy;", as this is doing the initialization atomic!
// (in opposite to lazy_static)
// this avoids that concurrent dart calls create a race condition
static API: Lazy<RwLock<AppState>> =
    Lazy::new(|| RwLock::new(AppState::new()));

// holds the complete state of the app, as a global static variable
#[derive(Default, Serialize, Deserialize, Debug)]
struct AppState{
    model: TodoListModel,
}

impl AppState {
    fn new() -> Self {
        //configures the logger
        OsLogger::new("com.example.todo_app")
        .level_filter(LevelFilter::Trace)
        .category_level_filter("Settings", LevelFilter::Trace)
        .init()
        .unwrap();
        debug!("initializing the Rust lib");
        // TODO handle error cases
        let app_state = AppState::load().unwrap();
        info!("Initialization finished, log level is {:?}", log::max_level());
        app_state
    }
}
```

@AlienKevin thanks for the solutions (an questions) you provided in the past!

**Reply by AlienKevin** [reply]
@patmuk Really appreciate your detailed solution. For convenience of migrating to V2 and ease of working with Future<Api> on the Dart side, I'm trying to use @fzyzcjy 's suggestion with the `RustOpaque`. However, I bumped into this error after marking my `Api` struct as opaque:
```
2023/12/18 09:30:18 [ERROR] panicked at 'not yet implemented: generate code for type "RichDict"', /Users/kevin/.cargo/registry/src/index.crates.io-6f17d22bba15001f/flutter_rust_bridge_codegen-1.82.1/src/ir/ty_unencodable.rs:59:9
```
And here's my `Api` struct, where the `RichDict` type is imported from an external crate and under the hood is just a type alias for `HashMap<usize, RichEntry>`.
```
#[frb(opaque)]
#[derive(Deserialize, Serialize)]
pub struct Api {
    dict: RichDict,
    #[serde(skip)]
    english_index: EnglishIndex,
    #[serde(skip)]
    pr_indices: Mutex<PrIndices>,
    #[serde(skip)]
    variants_map: VariantsMap,
    #[serde(skip)]
    word_list: HashMap<String, Vec<String>>,
}
```

Since I marked the whole `Api` as opaque, I don't quite understand why the `RichDict` needs to be encoded. I never exposed `RichDict` in the public interface of the `Api` struct.

**Reply by fzyzcjy** [reply]
Firstly, are you using `#[frb(opaque)]` on v1 or v2? It is v2-only feature IMHO.
For v1, just use `RustOpaque<WhateverType>` on every places you want to use the `WhateverType`.
I do expect v1 to contain some bug in the code generator (though they will not be bugs in production, but only fail to generate code)... :/ v2 code is refactored a lot to make it much cleaner imho.

**Reply by AlienKevin** [reply]
Oooh I see, was looking at the wrong doc. So now I wrapped all the fields of the struct with `RustOpaque`, the codegen finished successfully!
```rust
pub struct WordshkApi {
    dict: RustOpaque<RichDict>,
    english_index: RustOpaque<EnglishIndex>,
    pr_indices: RustOpaque<Mutex<PrIndices>>,
    variants_map: RustOpaque<VariantsMap>,
    word_list: RustOpaque<HashMap<String, Vec<String>>>,
}
```

Though I got a bunch of `cannot find type` errors when building Rust:
```
2023/12/18 09:56:31 [INFO] Phase: Running build_runner
2023/12/18 09:56:31 [INFO] Phase: Formatting Dart code
2023/12/18 09:56:31 [INFO] Success!
2023/12/18 09:56:31 [INFO] Now go and use it :)
[INFO  cargo_lipo::meta] Will build universal library for ["wordshk_api"]
[INFO  cargo_lipo::lipo] Building "wordshk_api" for "aarch64-apple-ios"
   Compiling wordshk_api v0.1.0 (/Volumes/Documents/personal/wordshk/rust)
error[E0412]: cannot find type `EnglishIndex` in this scope
   --> src/bridge_generated.io.rs:223:26
    |
223 | impl Wire2Api<RustOpaque<EnglishIndex>> for wire_EnglishIndex {
    |                          ^^^^^^^^^^^^ not found in this scope
    |
help: consider importing this type alias
    |
1   + use wordshk_tools::english_index::EnglishIndex;
    |
...
```

**Reply by AlienKevin** [reply]
If this won't be a problem in V2, I can also wait until V2 is stabalized and use the once_cell approach for now.

**Reply by fzyzcjy** [reply]
Hmm, I mean, wrap in function:`fn f(api: RustOpaque<WordApi>){}`

The auto-rust-opaque feature will not have this problem as well.

**Reply by AlienKevin** [reply]
Cool, I'll try out V2 when it's stablized. For now, the once_cell approach seems to be working quite well. I just have a separate `init_api` method on the Rust side that I always call first when accessing the api on the dart side:
```dart
late final myAppApi = MyAppApiImpl(dylib);
final apiInit = myAppApi.initApi();

Future<MyAppApiImpl> api() async {
  await apiInit;
  return myAppApi;
}
```

**Comment by fzyzcjy** [comment]
P.S. https://github.com/fzyzcjy/flutter_rust_bridge/discussions/962 also discusses the state-in-Rust topic. (Cross-linking for easier searching)
